salt/utils/asynchronous.py

```diff
@@ -2,95 +2,317 @@
+            self._ioloop._salt_started_called = True
+            self._ioloop.start()
         else:
-            return attr
+            log.warn("Tried to start event loop again: %s", stack())
 
-    def _block_future(self, future):
-        self.io_loop.add_future(future, lambda future: self.io_loop.stop())
-        self.io_loop.start()
-        return future.result()
+    def stop(self, *args, **kwargs):
+        self._ioloop._salt_started_called = False
+        self._ioloop.stop()
 
-    def __del__(self):
+    def close(self, *args, **kwargs):
+        log.debug("IOLoop.close called %s", stack())
+
+    def real_close(self, *args, **kwargs):
+        self._ioloop._salt_close_called = True
+        self._ioloop.close()
+        if self.sync_runner:
+            self.sync_runner.io_loop.close()
+            self.sync_runner = None
+
+    def run_sync(self, func, timeout=None):
+        if HAS_ASYNCIO:
+            asyncio_loop = asyncio.get_event_loop()
+        else:
+            asyncio_loop = False
+        if self.is_running() or (asyncio_loop and asyncio_loop.is_running()):
+            log.debug("run_sync - with running loop")
+            if self.sync_runner is None:
+                self.sync_runner = self.sync_runner_cls()
+            return self.sync_runner.run_sync(func)
+        else:
+            log.debug("run_sync - without running loop")
+            return self._ioloop.run_sync(func)
+
+    def is_running(self):
+        if USES_ASYNCIO:
+            try:
+                return self._ioloop.is_running()
+            except AttributeError:
+                pass
+            return self._ioloop.asyncio_loop.is_running()
+        else:
+            return self._ioloop._running
+
+
+try:
+    if IOLoop.seed_loop is None:
+        IOLoop.seed_loop = IOLoop()
+except AttributeError as exc:
+    # When running under sphinx builds the tornaod stuff is mocked so an
+    # attribute error is raised. Under normal operation this won't happen.
+    log.warn('IOLoop attribute error encountered out of context')
+
+
+class SyncWrapper(object):
+
+    def __init__(self, cls, args=None, kwargs=None, async_methods=None,
+            close_methods=None, loop_kwarg=None):
+        self.io_loop = tornado.ioloop.IOLoop()
+        if args is None:
+            args = []
+        if kwargs is None:
+            kwargs = {}
+        if async_methods is None:
+            async_methods = []
+        if close_methods is None:
+            close_methods = []
+        self.loop_kwarg = loop_kwarg
+        self.cls = cls
+        if loop_kwarg:
+            kwargs[self.loop_kwarg] = self.io_loop
+        self.obj = cls(*args, **kwargs)
+        self._async_methods = list(
+            set(async_methods + getattr(self.obj, 'async_methods', []))
+        )
+        self._close_methods = list(
+            set(close_methods + getattr(self.obj, 'close_methods', []))
+        )
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self):
```

Comment:
diito:
https://docs.quantifiedcode.com/python-anti-patterns/correctness/exit_must_accept_three_arguments.html

Reply:
Yep, I rushed through that and forgot about the `__exit__` signature.